contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,492 @@
+   recommendation provided by the VPA Recommender. If the Recommender is not
+   available, it falls back to the recommendation cached in the VPA object.
+   
+6. Pods can use the [**Initializer**](https://github.com/kubernetes/kubernetes/issues/3585)
+   mechanism to ensure they are not scheduled until a matching VPA object is
+   present. This is a recommended way of using VPA.
+   
+7. **VPA Updater** is a component responsible for **real-time updates** of Pods.
+   If a Pod uses VPA in `"Auto"` mode, the Updater can decide to update it with
+   recommender resources.
+   In MVP this is realized by just evicting the Pod in order to have it
+   recreated with new resources. This approach requires the Pod to belong to a
+   Replica Set (or some other owner capable of recreating it).
+   In future the Updater will take advantage of in-place updates, which would
+   most likely lift this constraint.
+   Because restarting/rescheduling Pods is disruptive to the service, it must be
+   rare.
+   
+8. VPA only controls the resource **request** of containers. It sets the limit
+   to infinity. The request is calculated based on analysis of the current and
+   previous runs (see [Recommendation model](#recommendation-model) below).
+   
+9. **History Storage** is a component that consumes utilization signals and OOMs
+   (same data as the Recommender) from Heapster and stores it persistently.
+   It is used by the Recommender to **initialize its state on startup**.
+   First implementation will be InFluxDB.
+   
+### Architecture overview ###
+![VPA Architecture Diagram](images/vpa-architecture.png "VPA architecture overview")
+
+### API ###
+We introduce a new type of API object `VerticalPodAutoscaler`, which
+consists of the [label selector](https://kubernetes.io/docs/api-reference/v1.5/#labelselector-unversioned)
+for matching Pods and two policy sections: the update policy and the resources
+policy.
+
+#### Example VPA object ####
+```json
+"target": {
+  "matchLabels": [{
+    "key": "app",
+    "value": "my-nginx-1",
+  }]
+}
+"updatePolicy" {
+  "mode": "Auto",
+}
+"resourcesPolicy" {
+  "containers": [{
+    "name": "my-nginx",
+    "requests": {
+      "cpu": {
+        "min": "500m",
+      },
+      "memory": {
+        "min": "200Mi",
+      }
+    }
+  }]
+}
+"recommendation" {
+  "containers": [{
+    "name": "my-nginx",
+    "resources": {
+      "requests": {
+        "cpu": "1200m",
+        "memory": "450Mi",
+      }
+    }
+  }]
+}
+```
+
+#### Label Selector ####
+The label selector determines which Pods will be scaled according to the given
+VPA policy. The Recommender will aggregate signals for all Pods matched by a
+given VPA, so it is important that the user set labels to group similarly
+behaving Pods under one VPA.
+
+It is yet to be determined how to resolve conflicts, i.e. when the Pod is
+matched by more than one VPA (this is not a VPA-specific problem though).
```

Comment:
The solution for everything else is to use controller ref.  Sounds like something like that should be used here.